### ISSUE TITLE
Analyzers incorrectly cache rule parameters from the SonarLint.xml file

### ISSUE THREAD

**Issue by duncanp-sonar** [SonarSource/sonar-dotnet#3361]
### Description

Rule parameters are passed to the analyzers via a file called `SonarLint.xml` using the Roslyn `AdditionalFiles` mechanism. Roslyn passes the additional files as part of the analysis context each time an analysis is triggered. The contents of the additional files can change between analysis runs.

The sonar-dotnet analyzers load the settings once per AppDomain and then ignore any settings that are supplied in subsequent analysis runs in the same AppDomain. This can lead to the wrong rule parameters being applied.

### Repro steps

1. Open a project in VS and install the SonarC# analyzer (or use SonarLint).
2. Add a `SonarLint.xml` settings with the content given below, with the `Build Action` set to `C# analyzer additional file` (or set the item type to `AdditionalFiles` if you are editing the project file directly.
3. Edit the ruleset to turn on rule S1451
4. Edit a code file to add a copyright header with the wrong year, and check the issue is raised.
4. Change the year in the copyright in the `SonarLint.xml` file so the rule should not fire, and save the file.
5. Edit the code to trigger the rule again.

### Expected behaviour
The new settings are picked up and the issue is not raised.

### Actual behaviour
The new settings are ignored until VS is restarted.

#### Repro SonarLint.xml file
```
<?xml version="1.0" encoding="UTF-8"?>
<AnalysisInput>
  <Settings/>
  <Rules>
    <Rule>
      <Key>S1451</Key>
      <Parameters>
        <Parameter>
          <Key>headerFormat</Key>
          <Value>// Copyright ACME 2020
</Value>
        </Parameter>
        <Parameter>
          <Key>isRegularExpression</Key>
          <Value>false</Value>
        </Parameter>
      </Parameters>
    </Rule>
  </Rules>
  <Files>
  </Files>
</AnalysisInput>
```

### Known workarounds

Restart VS.

### Related information

* C#/VB.NET Plugins Version: all versions with parameterised rules
* Visual Studio Version: all

### Background
The parameters are loaded and cached by the static `ParameterLoader` class that holds a static list of the parameterised analyzers it has processed. If it has already set the parameters for a particular analyzer it won't set them again - [see here](https://github.com/SonarSource/sonar-dotnet/blob/25c1ce8c65d14f2d284657a920dfe2490a3a93b7/sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/ParameterLoader.cs#L88).

In VS, this means that the once the analyzers have loaded a `SonarLint.xml` file, those rule parameters will be used until VS is restarted. It doesn't matter if the user edits the settings file, switches to a solution with a different `SonarLint.xml` file, or even a different solution that does not have a `SonarLint.xml` file.

I haven't checked whether command line builds are impacted, but my guess is that they are. MSBuild can reuse nodes to build subsequent projects, which might have different (or no) `SonarLint.xml` settings.

Also, both `dotnet` and `MSBuild` have shared build server processes that keep running after a build is complete that can then be used when building other solutions.
